Fix return type of setInterval in experimental/timers package

mkadirtan opened the pull request:
setInterval function returns the interval id, It is also stated in the jsdoc above the function. It was wrongly typed as void which causes type errors.

Please fill in this template.

- [+] Use a meaningful title for the pull request. Include the name of the package modified.
- [-] Test the change in your own code. (Compile and run.)
- [-] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [+] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [+] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [-] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [+] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [-] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

ppcano commented:
HI @mkadirtan, thanks for this fix.  Please kindly include a test on [test/timers.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/k6/test/timers.ts).

mstoykov commented on types/k6/experimental/timers.d.ts:
```diff
@@ -27,7 +27,7 @@ export function clearTimeout(timeoutID: TimeoutID): void;
  * @returns The interval id.
  */
-export function setInterval(functionRef: (...args: any[]) => void, delay: number, ...args: any[]): void;
+export function setInterval(functionRef: (...args: any[]) => void, delay: number, ...args: any[]): number;
```
```suggestion
export function setInterval(functionRef: (...args: any[]) => void, delay: number, ...args: any[]): IntervalID;
```

it not only is a number, but it specifically an IntervalID and only it can be used in the `clearInterval` call which is likely why you want this change in the first place

sheetalkamat reviewed (changes requested):
Fix per feedback

mkadirtan commented:
> It has been more than two weeks and this PR still has no reviews.
> 
> I'll bump it to the DT maintainer queue. Thank you for your patience, @mkadirtan.
> 
> (Ping @na--, @mstoykov, @codebien, @imiric, @oleiade, @olegbespalov, @ppcano, @nicolevanderhoeven.)

I fixed the IntervalID type, currently I am adding tests

mkadirtan commented:
> > It has been more than two weeks and this PR still has no reviews.
> > I'll bump it to the DT maintainer queue. Thank you for your patience, @mkadirtan.
> > (Ping @na--, @mstoykov, @codebien, @imiric, @oleiade, @olegbespalov, @ppcano, @nicolevanderhoeven.)
> 
> I fixed the IntervalID type, currently I am adding tests

I've added tests. Can you review again please

mstoykov commented on types/k6/test/timers.ts:
```diff
@@ -33,7 +33,16 @@ setInterval('functionRef', 5);
+
 // clearInterval
 
+// @ts-expect-no-error
```
```suggestion
```
This shouldn't error

mstoykov replied:
```suggestion
```

same as in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/63961/files#r1112686419

mstoykov commented on types/k6/test/timers.ts:
```diff
@@ -33,7 +33,16 @@ setInterval('functionRef', 5);
 // @ts-expect-error
 setInterval('functionRef', 5, 'arg');
 
+// @ts-expect-no-error
```
```suggestion
```

I don't think this exist as a syntax, when you want a none error you just leave it

mstoykov commented on types/k6/test/timers.ts:
```diff
@@ -33,7 +33,16 @@ setInterval('functionRef', 5);
+// @ts-expect-no-error
+clearInterval(intervalId)
+
+// @ts-expect-no-error
```
```suggestion
// @ts-expect-error
```
Here we do not use the result of a setInterval so we *do* expect an error

mstoykov replied:
It will be nicer if it does cause an error as you shouldn't really ever use "just" a number you should always use a value you got from `setInterval` but it is okay - we can fix this when somebody who knows how to make that happen finds this ;)

mkadirtan replied:
> It will be nicer if it does cause an error as you shouldn't really ever use "just" a number you should always use a value you got from `setInterval` but it is okay - we can fix this when somebody who knows how to make that happen finds this ;)

What you want to achieve is called "Nominal Type", however typescript doesn't support it. There is a workaround as described here: https://dnlytras.com/blog/nominal-types . But I've not seen a project that makes use of it. This is the way things go with typescript 😕

mstoykov commented:
Ready to merge

mstoykov commented:
Ready to merge

mstoykov commented:
@sheetalkamat I think this is because you requested changes. Can you please approve this PR

mkadirtan commented:
@sheetalkamat ping

mkadirtan commented:
@sheetalkamat Please, I've been waiting

mkadirtan commented:
@mstoykov can we merge this regardless? There is no point for waiting

mstoykov commented:
Ready to merge

mkadirtan commented:
@sheetalkamat @sheetalkamat @sheetalkamat

mstoykov commented:
Ready to merge

mstoykov commented:
I have no idea what was the problem in the end but after reapproaving it seems to have worked :tada: 

Thanks for the PR @mkadirtan and sorry for the technical poblems and the long time this took :(

mkadirtan commented:
> I have no idea what was the problem in the end but after reapproaving it seems to have worked 🎉
> 
> Thanks for the PR @mkadirtan and sorry for the technical poblems and the long time this took :(

probably because I requested another review from @sheetalkamat and it disabled the change request flag, anyways thanks for merging!
